src/login/pam_systemd.c

```diff
@@ -627,6 +627,29 @@ static int apply_user_record_settings(pam_handle_t *handle, UserRecord *ur, bool
         return PAM_SUCCESS;
 }
 
+static int configure_runtime_directory(
+                pam_handle_t *handle,
+                UserRecord *ur,
+                const char *rt) {
+
+        int r;
+
+        assert(handle);
+        assert(ur);
+        assert(rt);
+
+        if (!validate_runtime_directory(handle, rt, ur->uid))
+                return PAM_SUCCESS;
```

Comment:
Why is it considered `PAM_SUCCESS` if validation fails?

Reply:
THis is to cover for su/sudo situations, i.e. when you change your UID but little else. This means logind will still track you as the same session that was originally used for login, and since a session is associated with a user, and a runtime directory to a user too it would mean your XDG_RUNTIME_DIR after the su/sudo transition might still point to where you came from, i.e. the originating user's directory. And we don't really want that, because in the typical case where you transition to root then any tool you then start might leave root-owned files in XDG_RUNTIME_DIR, and that sucks.

Our current approach is to simply not set XDG_RUNTIME_DIR in that case. i.e. let's set XDG_RUNTIME_DIR when its  owner and our new UID actually match, but if it doesn't, suppress the env var totally. And this is not considered a failure, but just a regular codepath.

Yeah su/sudo are weird. It transitions some stuff and doesn't transition some other stuff. I think the best approach to deal with it is to only do the absolute minimum to support it, and rather than trying to figure out what su/sudo really should mean for some subsystem turn it off entirely, and that's what we do here.

I'll add a comment making this clearer